### PR TITLE
Closes #151: Implement interface end to end connection view

### LINKED ISSUE #151: 3.3.0-beta2 can't hide patchpanels from Topology to get end2end connection
Before 3.3.0 it was possible to hide patchpanels and show "end2end connections" this now doesn't work anymore, or i don't know how to :)

### PR DESCRIPTION
**Pull request by tgrusendorf**
PR to add back in and end to end connection view, similar to what was implemented before.

I'm simplified the approach, rather than tracing cables, to use the same queryset and fields as NetBox already uses for the 'Interface Connections' view (host:8000/dcim/interface-connections/)

Users may want to show different combinations of links when selecting end-to-end interface connections (e.g., show wireless links but not cables), so I've added separate selection options for cables and wireless links, similar to what has already been done with Circuit Terminations and Power Feeds.

image

The current code assumes that there will be only one path destination of type 'interface' for each interface connection.  I believe this is a valid assumption as NetBox doesn't allow you to connect multiple cables to a single interface.

@mattieserver, I'd appreciate your comments.

### CODE REVIEW

**Comment by tgrusendorf** [comment]
Closes issue #151.

**Comment by mattieserver** [comment]
Hi,

It might be a good idea to rename to filter label to something like " show logical connections" as the previous implemention showed the full path of the cable. If I added the "cabling" role to the intermediary device roles it traced and added any extra device in the path that matches that role, so it added the full path for the connection.  Whereas this implemention is pure logical, skipping any intermidiate device.


Selecting both "show cables" and "show end-to-end interface connects" show 2 cables per actual connection. (Same with wireless links). This could be "confusing" for users.

**Comment by tgrusendorf** [comment]
I agree - we could also name the filter label just "Show Interface Connections" to align with the existing NetBox view, but I think 'Show Logical Connections' is clearer.

Any suggestion to reduce the confusion if a user selects both "Show Cables" and "Show Logical Connections"?  Maybe we just add some instructions to the "Use" section in README.

**Comment by mattieserver** [comment]
Yes, updating the readme + release notes will do i think.
I might add the option in the views.py if 'show_interfaces' is enabled the disable the 'show_cables' option. (or just return nothing since we currently have no way the send a error msg).

**Comment by tgrusendorf** [comment]
Thanks for the commit, I've now updated the readme and would appreciate your comments.

**Comment by dreng** [comment]
> Thanks for the commit, I've now updated the readme and would appreciate your comments.

I hope you appreciate my comment as well. :-)

Instead of writing in the README.md how not to use it, I would rather make sure that you can't use it wrong. Use a radio field or a choice field instead of two check boxes as the given choices should be mutually exclusive.

Btw., would you mind to update your fork and resolve the merge conflicts?

**Comment by tgrusendorf** [comment]
Rebased to resolve merge conflicts. I'll play with a radio button for cables vs logical interfaces and see how it looks, but we'll need to include an option for 'neither' in case the user wants to only show other things (e.g., just power or wireless).  

This week in my dev system I actually used it with both cables and logical connections turned on while I was working on cables patched through multiple patch panels and wanted an overlay of the end-to-end connections, so I'm wondering if that's a rare but valid use case. Maybe having a different edge style for logical connections would help.

**Comment by dreng** [comment]
@tgrusendorf 
I just set up a fresh VM with the current NetBox develop branch (v3.4.1-dev) and I fetched your pull request. I also imported the [NetBox demo data](https://github.com/netbox-community/netbox-demo-data). This is what I get when I just tick "Show Logical Connections" and leave everything else as it is:

> Internal Server Error: /plugins/netbox_topology_views/topology/
Traceback (most recent call last):
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/contrib/auth/mixins.py", line 109, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/mario/src/netbox-topology-views/netbox_topology_views/views.py", line 629, in get
    topo_data = get_topology_data(
  File "/home/mario/src/netbox-topology-views/netbox_topology_views/views.py", line 429, in get_topology_data
    termination_b = { "termination_name": destination.name, "termination_device_name": destination.device.name, "device_id": destination.device.id }
AttributeError: 'Module' object has no attribute 'name'
Internal Server Error: /plugins/netbox_topology_views/topology/
Traceback (most recent call last):
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/contrib/auth/mixins.py", line 109, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/home/mario/src/.venvs/netbox/lib/python3.9/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/mario/src/netbox-topology-views/netbox_topology_views/views.py", line 629, in get
    topo_data = get_topology_data(
  File "/home/mario/src/netbox-topology-views/netbox_topology_views/views.py", line 429, in get_topology_data
    termination_b = { "termination_name": destination.name, "termination_device_name": destination.device.name, "device_id": destination.device.id }
AttributeError: 'Module' object has no attribute 'name'

**Comment by tgrusendorf** [comment]
Thanks @dreng.  I believe you've exposed two problems.  One is an error in my branch that I'll fix today, the other appears to be invalid connections in the 3.4 Netbox demo data.  In the demo data it is showing some interfaces connected directly to modules (e.g. dmi01-akrom-rtr01 GigabitEthernet0/0/0) when the cable shows it's connected to a circuit.  Cable traces on these connections cause an error as well.

image

image

**Comment by dreng** [comment]
@tgrusendorf Thank you for continuing to work on this. I have replaced the demo data in my environment with real data to avoid such supposed errors in the future.
I tinkered around with this PR today and I think it's just a few steps away from being ready to be merged. Some thoughts about two points we discussed before:

1. The arrows look great for it's purpose. I found it difficult to seperate cables from logical connections though. I defined a dotted line with arrows and found it very satisfying. Have a look at the screenshots below. What do you thing about it?
2. I think that cables and logical connections are redundant if "connection" and "link peer" are the same. While the first of the following two scenarios is okay and can be useful, the second one isn't. So I would suggest dropping the logical connections in this case.

logicalconnok

logicalconnbad png

File: netbox_topology_views/views.py [hunk 1]
```diff
@@ -217,6 +219,12 @@ def create_edge(
         edge["dashes"] = [5, 5, 3, 3]
         title = "Power Connection"
 
+    elif interface is not None:
+        title = "Interface Connection"
+        edge["width"] = 3
```
Comment (dreng) [inline netbox_topology_views/views.py:224]:
```suggestion
        edge["width"] = 3
        edge["dashes"] = [1, 10, 1, 10]
```

**Comment by dreng** [comment]
Looks pretty complete now. Although it is a good idea to let the user decide if he want's to see redundant cables and logical connections, I'm not sure if this should be a default setting in the __init__.py rather than a checkbox like all the other options. But this should be addressed in a new issue in my opinion since it is a general question if we move all of the checkboxes to __init_.py. In my opinion all of this settings should belong to a saved filter and should be consolidated into one drop-down.

@mattieserver I think this is ready to be merged. What do you think?

File: README.md [hunk 2]
```diff
@@ -85,6 +85,9 @@ PLUGINS_CONFIG = {
 | ignore_cable_type        | []                                                                                                                                             | The cable types that you want to ignore in the views                                                                   |
 | preselected_tags         | []                                                                                                                                             | The name of tags you want to preload                                                                                   |
 | draw_default_layout      | False                                                                                                                                          | (bool) Set to True if you want to load draw the topology on the initial load (when you go to the topology plugin page) |
+| hide_single_cable_logical_conns      | False                                                                                                                                          | (bool) Set to True if you want to hide duplicate cables & interface connections |
```
Comment (dreng) [inline README.md:88]:
```suggestion
| hide_single_cable_logical_conns      | False                                                                                                                                          | (bool) Set to True if you want to hide duplicate cables & logical connections |
```

**Review by dreng — CHANGES_REQUESTED** [review]

**Comment by mattieserver** [comment]
I think we can merge this.
If we move the options/filters it should be in a new PR.

If we get feedback regarding the duplicate cables/logical connections we can also look to add a message with the toast.js that was added in another PR.

**Comment by tgrusendorf** [comment]
Thanks - wasn't expecting to wake up today to find this already merged.

I put the redundant connection / cable option in init.py because I didn't think it really belonged on the 'Filters' tab as it's more of a visualization option (I guess the same could be said of 'Save Coordinates' too).  In the future maybe an additional tab is warranted for display options.

I added the option to change the dash style for interface connections because I'm not really loving the dashed style TBH, but that's just personal preference.  I suppose that's something else that could go on a display options tab in the future.

Thanks for the feedback and testing throughout this PR.

**Pull request merged**